### PR TITLE
feat(node): use multiple versions of node, can run old and new toolch…

### PR DESCRIPTION
**Pull request by dymart**
…ains and default behaviour is not broken

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
Can only use one nodejs version for a workspace

## What is the new behavior?
Able to use multiple nodejs versions in the same workspace

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Hopefully backward compatible to make transitions easier

### CODE REVIEW

File: examples/multiple_node_versions/WORKSPACE [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    path = "../../../rules_nodejs",
+)
+
+load("@build_bazel_rules_nodejs//nodejs:repositories.bzl", "nodejs_register_toolchains", "rules_nodejs_dependencies")
+load("@build_bazel_rules_nodejs//:index.bzl", "node_repositories")
+
+# This just gives us bazel-skylib
+rules_nodejs_dependencies()
+
+# if changes are made to the order update the tests as the test_node_version_default uses this order
```
Comment (alexeagle) [inline examples/multiple_node_versions/WORKSPACE:27]: I'd say "The order matters because Bazel will provide the first registered toolchain when a rule asks Bazel to select it"

File: internal/node/node.bzl [hunk 2]
```diff
@@ -591,7 +593,8 @@ Predefined genrule variables are not supported in this context.
         default = Label("//internal/node:loader.js"),
         allow_single_file = True,
     ),
-    "_node": attr.label(
+    "node": attr.label(
+        # the default needs to change and it shouldn't be private so that it can be set by the user or a macro
```
Comment (alexeagle) [inline internal/node/node.bzl:597]: "shouldn't be private" is now out-of-date right?

Comment (alexeagle) [reply internal/node/node.bzl:597]: I think https://github.com/bazelbuild/rules_nodejs/issues/3124 is where we'll be forced to change the default

File: internal/node/node.bzl [hunk 3]
```diff
@@ -638,6 +641,7 @@ nodejs_binary_kwargs = {
     "implementation": _nodejs_binary_impl,
     "outputs": _NODEJS_EXECUTABLE_OUTPUTS,
     "toolchains": [
+        # what is the best way to resolve this
```
Comment (alexeagle) [inline internal/node/node.bzl:644]: is that a question for me?

**Review by alexeagle — COMMENTED** [review]
nice start!!

File: examples/multiple_node_versions/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,125 @@
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("@bazel_skylib//rules:diff_test.bzl", "diff_test")
+load(":defs.bzl", "my_nodejs")
+load("@build_bazel_rules_nodejs//:index.bzl", "nodejs_binary", "nodejs_test")
+
+# Trivial test fixture: a nodejs program that writes to a file
+write_file(
+    name = "js",
+    out = "some.js",
+    content = ["require('fs').writeFileSync(process.argv[2], 'stuff')"],
```
Comment (gregmagolan) [inline examples/multiple_node_versions/BUILD.bazel:10]: To keep from growing to much duplicate test code, lets move this all over to `e2e/core`. It looks like the additional of

```
    if ctx.attr.toolchain:
        toolchain = ctx.attr.toolchain[platform_common.ToolchainInfo].nodeinfo
    else:
        toolchain = ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo
```
& the new attr are the load bearing changes in `my_nodejs` rule.

I'd make those changes there, bring in multiple node versions in that WORKSPACE and then bring in a test case to test that you can specify a custom toolchain. You'll want to `select` the toolchain so that it still works on multiple platforms (since we won't be relying on bazel selecting the correct one).

Something like

```
my_nodejs(
    name = "run_15",
    out = "thing_15",
    entry_point = "version.js",
    toolchain = select({
        "@bazel_tools//src/conditions:linux_x86_64": "@node15_linux_amd64//:node_toolchain",
        "@bazel_tools//src/conditions:darwin": "@node15_darwin_amd64//:node_toolchain",
        "@bazel_tools//src/conditions:windows": "@node15_windows_amd64//:node_toolchain",
    }),               
)
```
(doesn't cover the M1 case but on an M1 you can run an intel binary in rosetta compat mode)

File: internal/node/node.bzl [hunk 5]
```diff
@@ -248,7 +248,9 @@ fi
     #
     # Rules such as nodejs_image should use only ctx.toolchains["@build_bazel_rules_nodejs//toolchains/node:toolchain_type"].nodeinfo
     # when building the image as that will reflect the selected --platform.
-    node_tool_files = ctx.files._node[:]
+    node_tool_files = ctx.files.node[:]
```
Comment (gregmagolan) [inline internal/node/node.bzl:251]: The changes to node.bzl are unrelated to the basic e2e test case above right?

Lets do these in a separate PR where we add the `toolchain` attribute.

**Review by gregmagolan — APPROVED** [review]

File: examples/multiple_node_versions/WORKSPACE [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# http_archive(
+#     name = "rules_nodejs",
+#     sha256 = "a54a01b883beff66da057a1a9afc3be83de566d8ba78f5139f9b370026af5868",
+#     urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/4.4.0/rules_nodejs-core-4.4.0.tar.gz"],
+# )
+
+local_repository(
```
[the remark this answers is not in the document]
Comment (dymart) [reply examples/multiple_node_versions/WORKSPACE:11]: Have some questions about what the workflow looks like for this. Would someone normally use local repository when they are actively developing locally and then when they have a change switch it to http archive for the pr? Or do people normally run the integration test runner locally? Or is there something else?

File: examples/multiple_node_versions/BUILD.bazel (hunk 4 above)
Comment (alexeagle) [reply examples/multiple_node_versions/BUILD.bazel:10]: note that the select statement will sadly download toolchains for all three platforms, I don't think it can be helped but you should mention that in a comment here

File: internal/node/node.bzl (hunk 2 above)
Comment (dymart) [reply internal/node/node.bzl:597]: yes that is out of date, this should be address in a follow up pr that helps remove the reference to @nodejs

File: internal/node/node.bzl (hunk 3 above)
Comment (dymart) [reply internal/node/node.bzl:644]: Sorry that comment slipped through. It's more relevant to a follow up pr

File: internal/node/node.bzl (hunk 5 above)
Comment (dymart) [reply internal/node/node.bzl:251]: this change is still needed to include nodejs_binary and nodejs_test in this pr where node can be specified or not. A follow up pr will address the toolchain changes. This allows a user to set the node to use but doesn't fully remove the toolchain yet. Might be better to address this in the follow up pr

File: examples/multiple_node_versions/BUILD.bazel (hunk 4 above)
Comment (dymart) [reply examples/multiple_node_versions/BUILD.bazel:10]: Sounds good, I'll move those things over and add the select statements

**Comment by dymart** [comment]
The nodejs_binary has been removed from core but it uses the new core toolchain so the rules_nodejs core package was added to the base workspace. I want to make sure this is alright to do during the transition to core
